### PR TITLE
    Fix some ThreadPool performance counters.

### PR DESCRIPTION
**Pull request by jaykrell**
Fix some ThreadPool performance counters.
https://bugzilla.xamarin.com/show_bug.cgi?id=41294

The bug is that some performance counters are always zero.
Because we do not update them.
This addresses two of them: Work items added and thread count.

There is an argument to just leave it not working or
remove the non-working counters, granted.

There is PerformanceCounter class and you can Increment instances,
however its metadata and some of the implementation is in System.dll.
Most of the code is in native.
ThreadPool is in mscorlib.dll, so can't use System.dll.
A few options exist to address this.
We could move and forward the type. We could duplicate and rename
and possibly subset the type.
Here we add a new icall very specifically for the scenario.

There is slight cost for the icall, and for the atomic increment.
Doing this using the existing managed way would have the same
atomic increment. Using any new managed would have similar motivation.

There is an attempt to measure but the results were unclear.

Note that the count of work items is not specifically for QueueUserWorkItem
as one might expect, but also for any async calls.

Fix a little bit of typos.

### CODE REVIEW

File: mcs/class/referencesource/mscorlib/system/threading/threadpool.cs [hunk 1]
```diff
@@ -644,7 +644,9 @@ internal void MarkThreadRequestSatisfied()
                 count = prev;
             }
         }
-
+#if MONO
+        internal static IntPtr THREADPOOL_WORKITEMS = InternalPerformanceCounter.GetImpl("Mono Threadpool", "Work Items Added");
```
Comment (jaykrell) [inline mcs/class/referencesource/mscorlib/system/threading/threadpool.cs]: Width? Doesn't look good here..

File: mcs/class/referencesource/mscorlib/system/threading/threadpool.cs [hunk 2]
```diff
@@ -1179,6 +1184,35 @@ internal void SetWaitObject(WaitHandle waitObject)
...
+    internal sealed class InternalPerformanceCounter
+    {
+        [MethodImplAttribute (MethodImplOptions.InternalCall)]
+        internal static extern long UpdateValue (IntPtr impl, bool do_incr, long value);
+
+        [MethodImplAttribute (MethodImplOptions.InternalCall)]
+        internal static extern IntPtr GetImpl (string category, string counter,
+                string instance, string machine, out int/*PerformanceCounterType*/ ctype, out bool custom);
+
+        [ReliabilityContract (Consistency.WillNotCorruptState, Cer.MayFail)]
```
Comment (jaykrell) [inline mcs/class/referencesource/mscorlib/system/threading/threadpool.cs]: Not sure the attributes are needed/correct.

Comment (jaykrell) [reply mcs/class/referencesource/mscorlib/system/threading/threadpool.cs]: (and success/failure dubious)

File: mono/metadata/threadpool-worker-default.c [hunk 3]
```diff
@@ -571,6 +571,10 @@ worker_try_create (void)
 		return FALSE;
 	}
 
+#ifndef DISABLE_PERFCOUNTERS
```
Comment (jaykrell) [inline mono/metadata/threadpool-worker-default.c]: Much nicer and more efficient on native side.

File: mcs/class/referencesource/mscorlib/system/threading/threadpool.cs [hunk 4]
```diff
@@ -1179,6 +1184,35 @@ internal void SetWaitObject(WaitHandle waitObject)
...
 #endif // !MONO
 
+#if MONO // Duplicate just enough of System.Diagnostics.PerformanceCounter
+         // for use within mscorlib.dll (vs. System.dll).
+         // Alternatively, move the real one to mscorlib.dll and type-forward from System.dll.
+
+    internal sealed class InternalPerformanceCounter
+    {
+        [MethodImplAttribute (MethodImplOptions.InternalCall)]
+        internal static extern long UpdateValue (IntPtr impl, bool do_incr, long value);
```
Comment (akoeplinger) [inline mcs/class/referencesource/mscorlib/system/threading/threadpool.cs]: can be private (which we don't explicitly specify in our code style)

File: mcs/class/referencesource/mscorlib/system/threading/threadpool.cs [hunk 5]
```diff
@@ -1179,6 +1184,35 @@ internal void SetWaitObject(WaitHandle waitObject)
...
+
+        [MethodImplAttribute (MethodImplOptions.InternalCall)]
+        internal static extern IntPtr GetImpl (string category, string counter,
```
Comment (akoeplinger) [inline mcs/class/referencesource/mscorlib/system/threading/threadpool.cs]: can be private

File: mcs/class/referencesource/mscorlib/system/threading/threadpool.cs [hunk 6]
```diff
@@ -1207,7 +1241,6 @@ static internal bool PerformWaitCallback()
             if (dispatcher != null)
                 return dispatcher (ThreadPoolWorkQueue.Dispatch);
 #endif
-
```
Comment (akoeplinger) [inline mcs/class/referencesource/mscorlib/system/threading/threadpool.cs]: unrelated change, please revert

File: mcs/class/referencesource/mscorlib/system/threading/threadpool.cs [hunk 7]
```diff
@@ -1179,6 +1184,35 @@ internal void SetWaitObject(WaitHandle waitObject)
...
+        internal static extern long UpdateValue (IntPtr impl, bool do_incr, long value);
+
+        [MethodImplAttribute (MethodImplOptions.InternalCall)]
+        internal static extern IntPtr GetImpl (string category, string counter,
+                string instance, string machine, out int/*PerformanceCounterType*/ ctype, out bool custom);
+
+        [ReliabilityContract (Consistency.WillNotCorruptState, Cer.MayFail)]
+        internal static void Increment (IntPtr impl)
+        {
+            UpdateValue(impl, true, 1);
```
Comment (akoeplinger) [inline mcs/class/referencesource/mscorlib/system/threading/threadpool.cs]: is this thread safe in native?

File: mcs/class/referencesource/mscorlib/system/threading/threadpool.cs [hunk 8]
```diff
@@ -1179,6 +1184,35 @@ internal void SetWaitObject(WaitHandle waitObject)
     }
 #endif // !MONO
 
+#if MONO // Duplicate just enough of System.Diagnostics.PerformanceCounter
+         // for use within mscorlib.dll (vs. System.dll).
+         // Alternatively, move the real one to mscorlib.dll and type-forward from System.dll.
+
+    internal sealed class InternalPerformanceCounter
```
Comment (akoeplinger) [inline mcs/class/referencesource/mscorlib/system/threading/threadpool.cs]: I'd prefer not exposing the IntPtr for the perf counter, something like:

```csharp
internal sealed class InternalPerformanceCounter {
   IntPtr counter;
   ....

   public InternalPerformanceCounter (string categoryName, string counterName) {
       counter = GetImpl (categoryName, counterName, ...);
   }

   public void Increment () {
      UpdateValue (counter, true, 1);
   }
}
```

File: mcs/class/referencesource/mscorlib/system/threading/threadpool.cs (hunk 1 above)
Comment (akoeplinger) [reply mcs/class/referencesource/mscorlib/system/threading/threadpool.cs]: And now that I think about it, since we have DISABLE_PERFCOUNTERS on the native side this would crash when perfcounters are disabled.

File: mcs/class/referencesource/mscorlib/system/threading/threadpool.cs [hunk 9]
```diff
@@ -1145,10 +1149,11 @@ private bool ValidHandle()
 
     [System.Runtime.InteropServices.ComVisible(true)]
 #if FEATURE_REMOTING
-    public sealed class RegisteredWaitHandle : MarshalByRefObject {
+    public sealed class RegisteredWaitHandle : MarshalByRefObject
 #else // FEATURE_REMOTING
-    public sealed class RegisteredWaitHandle {
+    public sealed class RegisteredWaitHandle
 #endif // FEATURE_REMOTING
+    {
```
Comment (marek-safar) [inline mcs/class/referencesource/mscorlib/system/threading/threadpool.cs]: anything under `mcs/class/referencesource` flows from .net framework referencesource and better not to changes it to make the merges less painful

File: mcs/class/referencesource/mscorlib/system/threading/threadpool.cs (hunk 8 above)
Comment (marek-safar) [inline mcs/class/referencesource/mscorlib/system/threading/threadpool.cs]: Should be a static class

File: mono/metadata/icall-def.h [hunk 10]
```diff
@@ -896,6 +896,13 @@ ICALL(ILOCK_20, "Increment(int&)", ves_icall_System_Threading_Interlocked_Increm
 ICALL(ILOCK_21, "Increment(long&)", ves_icall_System_Threading_Interlocked_Increment_Long)
 ICALL(ILOCK_22, "Read(long&)", ves_icall_System_Threading_Interlocked_Read_Long)
 
+// A subset of System.Diagnostics.PerformanceCounter usable in mscorlib.
+// Note that System.Diagnostics.PerformanceCounter is mostly native,
+// but the metadata and small amount of implementation is in System.dll.
+ICALL_TYPE(THREADP_PERFCTR, "System.Threading.InternalPerformanceCounter", THREADP_PERFCTR_1)
+ICALL(THREADP_PERFCTR_1, "GetImpl", mono_perfcounter_get_impl)
+ICALL(THREADP_PERFCTR_2, "UpdateValue", mono_perfcounter_update_value)
```
[the remark this answers is not in the document]
Comment (lambdageek) [reply mono/metadata/icall-def.h]: Why is this needed? We already expose `mono_perfcounter_update_value` as `System.Diagnostics.PerformanceCounter:UpdateValue`.

As an aside: multiple icalls to the same C function is sometimes asking for trouble: people (e.g. me) sometimes forget to update both managed call sites when changing the native implementation because it's not obvious that there are multiple managed declarations.

**Comment by luhenry** [comment]
Also, as part of the Pull Request, please set an appropriate tittle and possibly a comment that explains what it is changing and why is the change.

**Comment by jaykrell** [comment]
From talking to Ludovic, I think probably won't take _quite_ this.

File: mono/metadata/threadpool-worker-default.c (hunk 3 above)
Comment (kumpera) [reply mono/metadata/threadpool-worker-default.c]: CAS on managed is equally efficient.

File: mcs/class/referencesource/mscorlib/system/threading/threadpool.cs [hunk 11]
```diff
@@ -675,7 +677,9 @@ public void Enqueue(IThreadPoolWorkItem callback, bool forceGlobal)
                     }
                 }
             }
-
+#if MONO
+            InternalPerformanceCounter.Increment(THREADPOOL_WORKITEMS);
```
Comment (kumpera) [inline mcs/class/referencesource/mscorlib/system/threading/threadpool.cs]: This is an icall on a fast path, no reason to do it.

Comment (kumpera) [reply mcs/class/referencesource/mscorlib/system/threading/threadpool.cs]: This will force a bucket of stuff to be brought in even when we don't need them.

File: mcs/class/referencesource/mscorlib/system/threading/threadpool.cs [hunk 12]
```diff
@@ -1145,10 +1149,11 @@ private bool ValidHandle()
 
     [System.Runtime.InteropServices.ComVisible(true)]
 #if FEATURE_REMOTING
-    public sealed class RegisteredWaitHandle : MarshalByRefObject {
+    public sealed class RegisteredWaitHandle : MarshalByRefObject
```
Comment (kumpera) [inline mcs/class/referencesource/mscorlib/system/threading/threadpool.cs]: Formating damage

**Comment by luhenry** [comment]
@kumpera if it's in managed, it means we would need to get `System.Diagnostics.PerformanceCounter` to be accessible from mscorlib, either by having it there and typeforward it from System to mscorlib, or by having an internal implementation in mscorlib which PerformanceCounter solely relies on. It seems like a big price to pay for a feature that not that many people depends on.

File: mcs/class/referencesource/mscorlib/system/threading/threadpool.cs (hunk 12 above)
Comment (jaykrell) [reply mcs/class/referencesource/mscorlib/system/threading/threadpool.cs]: The braces don't match up in editor unless it understands if.

File: mono/metadata/icall-def.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (jaykrell) [reply mono/metadata/icall-def.h]: Unclear if to wrap in ifdef -- the managed code makes the calls unconditionally.

File: mcs/class/referencesource/mscorlib/system/threading/threadpool.cs (hunk 7 above)
Comment (jaykrell) [reply mcs/class/referencesource/mscorlib/system/threading/threadpool.cs]: Yes.

File: mcs/class/referencesource/mscorlib/system/threading/threadpool.cs (hunk 12 above)
Comment (jaykrell) [reply mcs/class/referencesource/mscorlib/system/threading/threadpool.cs]: Coreclr fixed this over a year ago, and much more. Perhaps we are overdue for an update, or perhaps coreclr isn't the reference and I should learn what is..

Comment (jaykrell) [reply mcs/class/referencesource/mscorlib/system/threading/threadpool.cs]: eh, it is still in the referencesource actually.

Comment (jaykrell) [reply mcs/class/referencesource/mscorlib/system/threading/threadpool.cs]: The reference source is more managed as well.

File: mcs/class/corlib/Test/System.Threading/ThreadPoolTest.cs [hunk 13]
```diff
@@ -255,5 +255,29 @@ public void AsyncLocalCapture ()
...
+				ThreadPool.QueueUserWorkItem (_ => {});
+
+			var workItems1 = workItems.NextValue();
+			var threads0 = threads.NextValue();
+
+			Console.WriteLine ("workItems0:{0} workItems1:{1}", workItems0, workItems1);
+			Console.WriteLine ("{0}:{1}", "threads", threads0);
+
+			Assert.IsTrue (workItems1 - workItems0 == N, "#1");
+			Assert.IsTrue (threads0 > 0, "#2");
```
Comment (luhenry) [inline mcs/class/corlib/Test/System.Threading/ThreadPoolTest.cs]: There is no guarantee the ThreadPool is going to launch a new thread, so asserting on it is flaky.

File: mcs/class/corlib/Test/System.Threading/icallperf.pl [hunk 14]
```diff
@@ -0,0 +1,84 @@
+# Attempt to analyze the performance before and after collecting
```
Comment (lambdageek) [inline mcs/class/corlib/Test/System.Threading/icallperf.pl]: There are hardcoded paths in here.

It's probably worth either:
1. Not checking in this file at all, since it's not runnable.
2. or, fix it up so that it can be used on a clean checkout of Mono (just to gather numbers without comparing).
3. or, leave it non-runnable but with a comment that it's not intended to be working.

File: mcs/class/corlib/Test/System.Threading/ThreadPoolTest.cs (hunk 13 above)
Comment (jaykrell) [reply mcs/class/corlib/Test/System.Threading/ThreadPoolTest.cs]: Can items be run directly on invoking thread, i.e. on lightly loaded system, not worth creating a thread, or always queued to a separate thread? I think always queued.

File: mcs/class/corlib/Test/System.Threading/icallperf.pl (hunk 14 above)
Comment (jaykrell) [reply mcs/class/corlib/Test/System.Threading/icallperf.pl]: I changed it to accept command lines to run, and iteration count, but defaults to what I was using.

**Comment by luhenry** [comment]
Build

**Comment by luhenry** [comment]
Whitelist

File: mcs/class/corlib/Test/System.Threading/ThreadPoolTest.cs (hunk 13 above)
Comment (luhenry) [reply mcs/class/corlib/Test/System.Threading/ThreadPoolTest.cs]: See https://github.com/mono/mono/pull/6118#pullrequestreview-80534889

**Comment by luhenry** [comment]
@monojenkins build

**Comment by luhenry** [comment]
@monojenkins build Linux x64 mcs

**Comment by luhenry** [comment]
@monojenkins build Linux x64 Interpreter

**Comment by luhenry** [comment]
@monojenkins build Linux AArch64

**Comment by luhenry** [comment]
@monojenkins build Linux ARMv5 soft float

**Comment by luhenry** [comment]
@monojenkins build Linux ARMv7 hard float

**Comment by luhenry** [comment]
@monojenkins build Linux ARMv7 hard float Interpreter

**Comment by luhenry** [comment]
@monojenkins build Linux i386

**Comment by luhenry** [comment]
@monojenkins build Linux x64

**Comment by luhenry** [comment]
@monojenkins build Linux x64 Acceptance Tests

**Comment by luhenry** [comment]
@monojenkins build Linux x64 FullAOT

**Comment by jaykrell** [comment]
build

**Comment by luhenry** [comment]
@monojenkins build Linux ARMv7 soft float

**Comment by luhenry** [comment]
@monojenkins build Linux x64 Acceptance Tests

**Comment by luhenry** [comment]
@monojenkins build Linux x64 FullAOT

**Comment by luhenry** [comment]
@monojenkins build Linux x64 Interpreter

**Comment by luhenry** [comment]
@monojenkins build Linux x64 mcs

**Comment by luhenry** [comment]
@monojenkins build Linux x64

**Comment by luhenry** [comment]
@monojenkins build Linux x64 Interpreter

**Comment by luhenry** [comment]
@monojenkins build Linux ARMv7 hard float

**Comment by luhenry** [comment]
@monojenkins build Windows x64

**Comment by marek-safar** [comment]
There is corlib test build error

`Test/System.Threading/ThreadPoolTest.cs(264,24): error CS0246: The type or namespace name 'PerformanceCounter' could not be found (are you missing a using directive or an assembly reference?)`

**Comment by jaykrell** [comment]
Hm, it was probably working before later change where I moved the test by request into existing tests. Let me recheck and fix..

**Comment by jaykrell** [comment]
@monojenkins Linux x64

**Comment by jaykrell** [comment]
@monojenkins build Linux x64
